consensus/src/consensus.rs

```diff
@@ -52,6 +53,68 @@ impl<PublicKey: VerifyingKey> ConsensusState<PublicKey> {
         }
     }
 
+    pub async fn new_from_store(
+        genesis: Vec<Certificate<PublicKey>>,
+        metrics: Arc<ConsensusMetrics>,
+        recover_last_committed: HashMap<PublicKey, Round>,
+        cert_store: Store<CertificateDigest, Certificate<PublicKey>>,
+        gc_depth: Round,
+    ) -> Self {
+        let last_committed_round = *recover_last_committed
+            .iter()
+            .max_by(|a, b| a.1.cmp(b.1))
+            .map(|(_k, v)| v)
+            .unwrap_or_else(|| &0);
+
+        if last_committed_round == 0 {
+            return Self::new(genesis, metrics);
+        }
+        metrics
+            .recovered_consensus_state
+            .with_label_values(&[])
+            .inc();
+
+        let dag =
+            Self::construct_dag_from_cert_store(cert_store, last_committed_round, gc_depth).await;
+
+        Self {
+            last_committed_round,
+            last_committed: recover_last_committed,
+            dag,
+            metrics,
+        }
+    }
+
+    pub async fn construct_dag_from_cert_store(
+        cert_store: Store<CertificateDigest, Certificate<PublicKey>>,
+        last_committed_round: Round,
+        gc_depth: Round,
+    ) -> Dag<PublicKey> {
+        let mut dag: Dag<PublicKey> = HashMap::new();
+        let min_round = last_committed_round - gc_depth;
```

Comment:
Can this underflow?

Reply:
Very good eye! Indeed is possible and this has been fixed here https://github.com/MystenLabs/narwhal/pull/497